### PR TITLE
Dependency Injection

### PR DESCRIPTION
**Pull request by andrewcchau**
Refactored project to use dependency injection

### CODE REVIEW

File: src/main/java/lithium/university/services/TwitterService.java [hunk 1]
```diff
@@ -13,21 +15,22 @@
 import twitter4j.TwitterFactory;
 import twitter4j.conf.ConfigurationBuilder;
 
+import javax.inject.Inject;
+import javax.inject.Singleton;
 import java.util.List;
 import java.util.Optional;
 import java.util.stream.Collectors;
 
+@Module
```
Comment (kev-chan) [inline src/main/java/lithium/university/services/TwitterService.java]: you should not need @Module on TwitterService

File: src/main/java/lithium/university/services/TwitterService.java [hunk 2]
```diff
@@ -13,21 +15,22 @@
...
 public class TwitterService {
-    private static final TwitterService INSTANCE = new TwitterService();
     private final Logger logger = LoggerFactory.getLogger(TwitterService.class);
     private TwitterProperties twitterProperties;
 
-    private TwitterService() {
-    }
+    @Provides
+    @Singleton
+    public TwitterService provideTwitterService() { return new TwitterService(); }
```
Comment (kev-chan) [inline src/main/java/lithium/university/services/TwitterService.java]: this should not be needed

File: src/main/java/lithium/university/TwitterApplication.java [hunk 3]
```diff
@@ -1,15 +1,26 @@
...
+import javax.inject.Singleton;
 
 
 public class TwitterApplication extends Application<TwitterConfiguration> {
     public static final int TWEET_TOTAL = 25;
     public static final int TWEET_LENGTH = 280;
 
+    @Singleton
+    @Component(modules = TwitterService.class)
+    interface TwitterSetUp {
```
Comment (kev-chan) [inline src/main/java/lithium/university/TwitterApplication.java]: let's move it to a separate class

**Review by kev-chan — CHANGES_REQUESTED** [review]
create a Provider for Twitter

File: src/main/java/lithium/university/TwitterProvider.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import twitter4j.Twitter;
+import twitter4j.TwitterFactory;
+import twitter4j.conf.ConfigurationBuilder;
+
+import javax.inject.Provider;
+
+@Module
+public class TwitterProvider implements Provider {
```
Comment (kev-chan) [inline src/main/java/lithium/university/TwitterProvider.java]: this class can be simplified
do you need to implement Provider and use the @Provides annotation?

File: src/main/java/lithium/university/services/TwitterService.java [hunk 5]
```diff
@@ -10,24 +9,16 @@
...
     private final Logger logger = LoggerFactory.getLogger(TwitterService.class);
-    private TwitterProperties twitterProperties;
 
-    private TwitterService() {
-    }
-
-    public static TwitterService getInstance() {
-        return INSTANCE;
-    }
+    @Inject public TwitterService() {}
```
Comment (kev-chan) [inline src/main/java/lithium/university/services/TwitterService.java]: do you actually need @Inject?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/lithium/university/services/TwitterService.java (hunk 5 above)
Comment (andrewcchau) [reply src/main/java/lithium/university/services/TwitterService.java]: You'll get a compilation error "Dagger/MissingBinding" without @Inject

**Pull request merged**